crates/walrus-service/src/node/blob_sync.rs

```diff
@@ -462,10 +462,33 @@ impl BlobSynchronizer {
             .await
             .expect("database operations should not fail");
 
-        let futures_iter = this.node.owned_shards().into_iter().map(|shard| {
-            this.clone()
-                .recover_slivers_for_shard(shared_metadata.clone(), shard)
-        });
+        // Note that we only need to recover the slivers in the shards that are assigned to the
+        // node at the time of the blob certification. Due to slow event processing, the contract
+        // might have entered the new epoch with new shard assignment. Upon discovery of the new
+        // shard assignment, the node only needs to recover the slivers assigned in the epoch
+        // of blob certification due to that:
+        //   - The node may not have created the new shards yet.
+        //   - Since the blob is certified in an earlier epoch, it is this node's responsibility to
+        //     hold, recover, and transfer the shard to the new owner.
+        //
+        // If the node is severally lagging behind and the certified_epoch is 2 epochs older,
+        // this function panics since no shard assignment info is found. Upon restarting the node,
+        // the node will enter recovery mode until catching up with all the events and start
+        // recovering all the missing blobs.
+        let futures_iter = this
+            .node
+            .owned_shards_at_epoch(this.node.current_event_epoch())
```

Comment:
Hi @mlegner , simtest revealed an issue that the previous version of using this.certified_epoch (the event's epoch) here is actually incorrect. This is because there can be a high lag between the certified event is processed, and the start of blob recovery. So it is not guaranteed that when the blob sync task is scheduled to run, the node is still at certified_event.epoch.

The correct fix here is to use the epoch the latest event is in. And in order to do that, we need to track it. I added latest_event_epoch in StorageNodeInner, and updated it when encounter epoch change start event.

Let me know if you see any problem of current solution.

Reply:
Thanks a lot for catching this! 🙏 

Just wondering about a few corner cases:
* Is it a problem if the `current_event_epoch` is updated before we create new shard storage?
* What happens if we later lose one of the shards for which we now try to recover slivers?

I believe those are ok, but just want to make sure.

Reply:
Thanks for the review @mlegner !

> Is it a problem if the current_event_epoch is updated before we create new shard storage?

Yes, because after updating the current_event_epoch, the blob sync will use the new config, but we haven't created the storage yet.

> What happens if we later lose one of the shards for which we now try to recover slivers?

Recovering shards always use the "current" view of the walrus system. So shards cannot change when blob sync starts. However, are you saying if a blob sync takes so long that during it's syncing, a shard moves out, and then
deleted? Yeah, I think this is an issue, but only if a blob sync takes longer than 2 epoch duration, so which may be
rare.